### ISSUE TITLE
Feature request: configure secure app spawning per-app

### ISSUE THREAD

**Issue by Seb-135** [GrapheneOS/os-issue-tracker#2170]
I'm not certain if this is possible, but it would be nice to be able to turn off secure app spawning for each app (perhaps adding it as a third option in in Advanced > Configure Hardening in app info)
This would solve #1908 without needing to trick faulty anti-tamper such as LIAPP, and it (probably?) won't sacrifice the security provided by secure app spawning in general.
I have tested and it is possible to turn off secure app spawning, open a LIAPP-protected app, and then immediately turn it back on again. Perhaps the requested toggle could do something similar?

**Comment by repairmant** [comment]
I have only 1 app that needs this disabled to work really dont want to disable this system wide for one app

**Comment by davidfrickert** [comment]
This became more annoying since disabling secure app spawning now requires a reboot. I am in the same situation, only one app requires this setting to the disabled to work.

**Comment by thestinger** [comment]
@davidfrickert That change came along with another change avoiding the known compatibility issues it used to cause.

**Comment by davidfrickert** [comment]
> @davidfrickert That change came along with another change avoiding the known compatibility issues it used to cause.

Oh. It still causes issues with my app then 🥲

**Comment by thestinger** [comment]
Which app?

**Comment by davidfrickert** [comment]
> Which app?

SBI新生銀行

Version 3.8.0

com.shinseibank.powerdirect
versionCode 83

Its a japanese bank app. I can send you the APK if you'd like to check it. Its available on google play but region locked to Japan.

**Comment by thestinger** [comment]
Can you send the APK to logs@grapheneos.org?

**Comment by davidfrickert** [comment]
> Can you send the APK to logs@grapheneos.org?

Sent

**Comment by peterc-s** [comment]
For reference, another banking app has this same issue (won't work with secure app spawning) https://github.com/PrivSec-dev/banking-apps-compat-report/issues/341#issuecomment-3861492696

**Comment by ccalhoun1999** [comment]
It may have fixed itself in the past year, but Japan's eplus app (for concert tickets) used to throw errors when you open the actual ticket page in the app if secure app spawning is on.

**Comment by davidfrickert** [comment]
@thestinger did you have a chance to check this?

**Comment by davidfrickert** [comment]
I decompiled the app and it seems like the app has some root detection code that triggers when secure app spawning feature is used. Note: Claude used to analyze decompiled code.

> The app blocks specifically on GrapheneOS secure app spawning because libvoltmxrange.so (the HCL Volt MX security library) performs a Zygote spawn verification using raw syscall to read /proc process lineage, dl_iterate_phdr + mincore to verify pre-resident Zygote pages, and getauxval to check ELF auxiliary vector values — all of which are different when the OS spawns apps via fresh exec instead of fork() from Zygote.

**Comment by thestinger** [comment]
That's very strange. You should try reporting the issue to both the app developers and the makers of the library. GrapheneOS is much more secure rather than less secure. Exec-based spawning is a much more secure way to spawn apps where they don't share their memory tags, canaries, ASLR bases and more with every other app including many app-based base OS components such as system_server and many apps with privileges unavailable to regular apps. They should add support for GrapheneOS. They can still do these kinds of checks while permitting it. You can send them https://grapheneos.org/articles/attestation-compatibility-guide too even though they aren't currently using attestation in case they're planning to use it.

**Comment by davidfrickert** [comment]
> That's very strange. You should try reporting the issue to both the app developers and the makers of the library. GrapheneOS is much more secure rather than less secure. Exec-based spawning is a much more secure way to spawn apps where they don't share their memory tags, canaries, ASLR bases and more with every other app including many app-based base OS components such as system_server and many apps with privileges unavailable to regular apps. They should add support for GrapheneOS. They can still do these kinds of checks while permitting it. You can send them https://grapheneos.org/articles/attestation-compatibility-guide too even though they aren't currently using attestation in case they're planning to use it.

Sure, I will try. But I doubt they will care. Would be great if it could be possible to disable the secure app spawning per-app.

**Comment by thestinger** [comment]
@davidfrickert Disabling secure spawning for a specific app would force not using various other compatibility features since they're enabled in the Zygote process. Enabling those normally force enables secure spawning.

**Comment by davidfrickert** [comment]
> [@davidfrickert](https://github.com/davidfrickert) Disabling secure spawning for a specific app would force not using various other compatibility features since they're enabled in the Zygote process. Enabling those normally force enables secure spawning.

I think that's fine, at least for me, i would like to be able to disable secure app spawning + all other features that depend on it for a specific app and get my app to work, while keeping it enabled for all apps that don't have issues with it. Right now I'm forced to disable it globally if I want to use the banking app.

**Comment by thestinger** [comment]
A lot of those apps also need hardened_malloc and even the large address space disabled which won't be possible to disable without secure spawning since the Zygote has them enabled.

**Comment by peterc-s** [comment]
> A lot of those apps also need hardened_malloc and even the large address space disabled which won't be possible to disable without secure spawning since the Zygote has them enabled.

I think it'd still be preferable to be able to disable whatever can be disabled at a more granular level if it is possible to as it might resolve compatibility issues with some of these banking apps which have dev teams that simply don't care to support GOS without downgrading security globally.

**Comment by Jobe1105** [comment]
> > A lot of those apps also need hardened_malloc and even the large address space disabled which won't be possible to disable without secure spawning since the Zygote has them enabled.
> 
> I think it'd still be preferable to be able to disable whatever can be disabled at a more granular level if it is possible to as it might resolve compatibility issues with some of these banking apps which have dev teams that simply don't care to support GOS without downgrading security globally.

Agree in this case, @thestinger . I do understand the importance of the security features but if some essential apps need to be run then not having granular level controls would mean that we would need to disable security features for all apps. That is a larger security threat than at least disabling it for only certain apps. A lot of these dev teams simply do not care enough about GOS to implement exceptions regardless if we report this or not. 

Edit: perhaps it can be baked into the exploit compatibility settings to try to make apps work? Just thinking if we already have some granular level controls for exploit protection then why not add more? Just a suggestion.

**Comment by peterc-s** [comment]
> A lot of these dev teams simply do not care enough about GOS

Yeah, I know at least Barclays does not care and there are probably many other companies that have the same mentality. It seems a lot of people on the banking app compatibility report have been outright switching banks but that's not an ideal solution.

This probably isn't that high of a priority ticket and the GOS dev team themselves are likely stretched. Thanks for all you do (@thestinger and co.)

**Comment by e-t-l** [comment]
> A lot of those apps also need hardened_malloc and even the large address space disabled which won't be possible to disable without secure spawning since the Zygote has them enabled.

Are you saying that disabling Secure App Spawning prevents per-app toggling of hardened_malloc and extended virtual address space? 

Is there any way for the user to tell whether this is happening? 

For context, I have...
- **Secure App Spawning**: `Disabled`
- **hardened_malloc**: `Default (Enabled)`
- **Google Camera app**: hardened_malloc `Disabled`
- No indication in UI that any of these settings are conflicting or overridden

If I understand you correctly, then the Secure App Spawning setting should result in hardened_malloc being force-enabled in Google Camera. If this is the case, I would have expected something like the Google Camera app's compatibility settings to show certain compatibility toggles as grayed-out, with a message like "This setting cannot be changed while Secure App Spawning is disabled."

Otherwise, it APPEARS (from the UI) that all my settings are being applied without conflicts. 

Can you clarify whether certain settings are being applied or overridden?

**Comment by d34ndev** [comment]
Nice, looks like this feature has been added in the latest release - https://grapheneos.org/releases#2026071100
